### DISCUSSION TITLE
How can I make MDX work with i18n?

### DISCUSSION THREAD

**Discussion by wanku-x** [vercel/next.js#44885] [Help]
By default, in nextjs, one mdx file corresponds to several locales. But this is not suitable for me, because then I will have to use dictionaries. I would like to have different mdx files for each locale.

For example, something like this:

```
/pages
- /installation
-- /index.mdx        // Roure `/installation/`
-- /index.en.mdx     // Route `/en/installation/`
-- /guide
--- /index.mdx       // Route `/installation/guide/`
--- /index.en.mdx    // Route `/en/installation/guide/`
- /index.mdx         // Route `/`
- /index.en.mdx      // Route `/en/`
- /404.mdx           // Not found page for RU locale
- /404.en.mdx        // Not found page for EN locale
```

I tried to get started with the main page and implement it using rewrite in `next.config.js`:

```js
// next.config.js
...

const nextConfig = withMDX({
    reactStrictMode: true,
    i18n: {
        locales: ['en', 'ru'],
        defaultLocale: 'ru',
    },
    pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
    trailingSlash: true,
    async rewrites() {
        return {
            beforeFiles: [
                {
                    source: '/en/',
                    destination: '/index.en/',
                    locale: false,
                },
            ],
        }
    },
})
```

It works. But I will need to add other locales in the future. I tried adding the lang parameter:

```js
// next.config.js
...

const nextConfig = withMDX({
    reactStrictMode: true,
    i18n: {
        locales: ['en', 'ru'],
        defaultLocale: 'ru',
    },
    pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
    trailingSlash: true,
    async rewrites() {
        return {
            beforeFiles: [
                {
                    source: '/:lang([a-z]{2})/',
                    destination: '/index.:lang/',
                    locale: false,
                },
            ],
        }
    },
})
```

As a result, this led to the fact that `index.en.mdx` is actually displayed on the `/en/` page, but the `/` page returns 404 page.

I thought maybe regexp was not working properly. And maybe for the path `/` it is trying to find the file `index..mdx`. I read in the documentation that the [path-to-regexp](https://github.com/pillarjs/path-to-regexp) library is used for regular expressions in nextjs paths. I tried using it to check how the paths `/` and `/en/` are matched with the regular expression `/:lang([a-z]{2})/`. And on the path `/` it returned null, and on the path `/en/` it returned an array matching the regular expression.

That is, my assumption turned out to be incorrect. I postponed this problem for later, and decided to delete the `/index.en/` page, which I don't need. I tried to do it like this:

```js
// next.config.js
...

const nextConfig = withMDX({
    reactStrictMode: true,
    i18n: {
        locales: ['en', 'ru'],
        defaultLocale: 'ru',
    },
    pageExtensions: ['ts', 'tsx', 'js', 'jsx', 'md', 'mdx'],
    trailingSlash: true,
    async rewrites() {
        return {
            beforeFiles: [
                {
                    source: '/:lang([a-z]{2})/',
                    destination: '/index.:lang/',
                    locale: false,
                },
                {
                    source: '/index.:lang([a-z]{2})/',
                    destination: '/404/',
                },
            ],
        }
    },
})
```

The paths `/index.en/` and `/en/index.en/` began to return the 404 page as I wanted. But along with them, the pages `/` and `/en/` also began to return 404 pages. This discouraged me)

Okay, I deleted all the rewrites from the file `next.config.js`. I decided to try to set up routing via middleware.

```js
// middleware.js

import { NextResponse } from 'next/server'

export function middleware(request) {
    const { pathname, locale } = request.nextUrl

    if (
        pathname === '/' &&
        locale !== 'ru'
    ) {
        return NextResponse.rewrite(new URL(`/index.${locale}/`, request.url))
    }
}

export const config = {
    matcher: [
        '/',
        '/((?!api|_next|favicon.ico).*)',
    ],
}
```

The path `/` works fine, but switching to the path `/en/` opens the page first `index.en.mdx`, and then there seems to be a redirect on the client side to `/` (the path `/en/` returns code 200, so it doesn't look like it was a redirect on the server side).

As a result, I could not figure out how to implement my plans. Can someone suggest how this can be done?

Repo: https://github.com/wanku-x/next-mdx-i18n

**Comment by artt** [comment]
Looking to do something similar. Do you have any updates on this or how did you end up doing it?
